asf_public_discourse_home_decarbonisation/analysis/mse/eda_mse_category_data.py

```diff
@@ -0,0 +1,311 @@
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def add_datetime_variables(mse_data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Transform the datetime variable into a datetime object and add year and year_month variables.
+    Args:
+        mse_data (pd.DataFrame): dataframe with MSE data
+    Returns:
+        pd.DataFrame: Pandas DataFrame with datetime variables added
+    """
+    mse_data["datetime"] = mse_data["datetime"].apply(
+        lambda x: datetime.strptime(x, "%Y-%m-%dT%H:%M:%S+00:00")
+    )
+    mse_data["year"] = mse_data["datetime"].apply(lambda x: x.year)
+    mse_data["year_month"] = mse_data["datetime"].dt.to_period("M")
+
+    return mse_data
+
+
+def create_statistics_from_data(mse_data: pd.DataFrame):
+    """
+    Create and prints statistics from the MSE data.
+
+    Args:
+        mse_data (pd.DataFrame): _description_
+    """
+    if len(mse_data) == len(mse_data.drop_duplicates()):
+        logger.info("There are no duplicated posts in the data.")
+    else:
+        logger.warning(
+            f"There are duplicated posts in the data.\nNumber instances: {len(mse_data)}\nNumber unique instances: {len(mse_data.drop_duplicates())}"
+        )
+
+    logger.info(f"Categories in data: {mse_data['category'].unique()}")
+
+    logger.info(
+        f"Number of instances (posts & replies) collected for category '{category}': {len(mse_data)}"
+    )
+
+    number_original_posts = len(mse_data[mse_data["is_original_post"] == 1])
+    logger.info(
+        f"Number of original posts collected for category '{category}': {number_original_posts}"
+    )
+
+    logger.info(
+        f"Data collected ranges from {mse_data['datetime'].min()} to {mse_data['datetime'].max()}"
+    )
+
+
+def plot_number_posts_per_year(mse_data: pd.DataFrame, category: str):
+    """
+    Plots and saves a bar plot with number of posts per year.
+
+    Args:
+        mse_data (pd.DataFrame): Money Saving Expert data
+        category (str): MSE category/sub-forum
+    """
+    posts_per_year = mse_data.groupby("year")[["id"]].nunique()
+    posts_per_year.columns = ["number_posts"]
+    posts_per_year.reset_index(inplace=True)
+
+    chart = (
+        alt.Chart(posts_per_year)
+        .mark_bar()
+        .encode(
+            y=alt.Y("year:O", title="Year"),
+            x=alt.X("number_posts:Q", title="Number of posts"),
+            color=alt.value(pc.NESTA_COLOURS[1]),
+        )
+        .properties(width=400, height=300)
+        .interactive()
+    )
+    chart = pc.configure_plots(chart, "Number of posts per year", "")
+    path_to_plot = f"{MSE_FIGURES_PATH}/category_{category}_posts_per_year.html"
+    chart.save(path_to_plot)
+
+    logger.info(
+        f"Plot with number of posts per year for category '{category}' saved successfully in {path_to_plot}"
+    )
+
+
+def check_username_id_correspondence(mse_data: pd.DataFrame):
+    """
+    Checks if there is a 1-1 correspondence between `username` and `user_id`.
```

Comment:
Not entirely sure that I understand what this function used for in a broader sense. Would a username not just have a single user_id? It seems not, but then why is that the case?

Reply:
When users delete their accounts then their username will appear as "Deleted User", but the user_id will always survive. I explored this initially, just to check these two variables were showing the same info! :)